// @slicemachine/adapter-next/dist/hooks/slice-asset-read.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fs = require("node:fs/promises");
const buildSliceAssetPath = require("../lib/buildSliceAssetPath.cjs");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const fs__namespace = /* @__PURE__ */ _interopNamespaceDefault(fs);
const sliceAssetRead = async (data, context) => {
  const filePath = await buildSliceAssetPath.buildSliceAssetPath({
    libraryID: data.libraryID,
    sliceID: data.sliceID,
    assetID: data.assetID,
    helpers: context.helpers,
    actions: context.actions
  });
  const assetData = await fs__namespace.readFile(filePath);
  return {
    data: assetData
  };
};
exports.sliceAssetRead = sliceAssetRead;
